projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/types/FunctionType.java

```diff
@@ -55,6 +57,11 @@ public Type getReturnType() {
         return returnType;
     }
 
+    public void setReturnType(Type returnType) {
+        CompilerAsserts.neverPartOfCompilation();
```

Comment:
You have to invalidate all compilation units that depend on this field (similar to what we do in PointerType).
If possible, make returnType final. 
If not, you have to use an assumption (see PointerType) to make this save.